### DISCUSSION TITLE
Separating commits by branch

### DISCUSSION THREAD

**Discussion by regexaurus** [community/community#22889] [New to GitHub]
<p>I’m not exactly new to git/github, but I’m not a regular user/contributer and still struggle with the fundamentals. Today, I made a minor typo correction to regexaurus/linkchecker (a fork of linkchecker/linkchecker), on branch <em>typos-regexaurus</em>, committed the change, pushed the commit/branch and opened a PR.</p>
<p>I later checked out a new branch, <em>issueurl-regexaurus</em>, made an edit that should satisfy issue <span class="hashtag">#174</span> on linkchecker/linkchecker, committed the change, pushed the commit/branch and opened a PR. I was surprised to see that the second PR included/showed the earlier commit as well.</p>
<p>I don’t think this is desirable, and expected only the second/related commit to appear in the second PR.</p>
<p>I would appreciate suggestions on workflow to produce my expected outcome, or help understanding why the actual outcome is normal/preferred.</p>
<p>Thank you!</p>

**Comment by imasaru** [comment, marked as answer]
<blockquote>
<p><a class="mention" href="/regexaurus">@regexaurus</a> wrote:</p>
</blockquote>
<blockquote>
<p>I later checked out a new branch, <em>issueurl-regexaurus</em>, made an edit that should satisfy issue <span class="hashtag">#174</span> on linkchecker/linkchecker, committed the change, pushed the commit/branch and opened a PR. I was surprised to see that the second PR included/showed the earlier commit as well.</p>
</blockquote>
<blockquote>
<p>I don’t think this is desirable, and expected only the second/related commit to appear in the second PR.</p>
</blockquote>
<p>You are right, that is not desirable. You made a PR to address a single issue and its changes should fix nothing but that issue :smiley:</p>
<p>It sounds to me like you may have created <em>issueurl-regexaurus</em> from <em>typos-regexaurus.</em> You may have been on the page of your fork, with the latter branch selected for viewing, and then created the former from the <strong>Branch</strong> dropdown. Instead, you should have selected the branch that is even with upstream/parent and created the former from <em>it.</em></p>
<p>To fix this, I would (personally) clone <em>issueurl-regexaurus</em> to my machine,  <strong>git reset --hard</strong>  to be even with <em>linkchecker/linkchecker:master</em>, and then <strong>git cherry-pick <a href="https://github.com/linkchecker/linkchecker/pull/176/commits/5a08a55a30b10c84b41cd050fcfcb12254032ca1" rel="nofollow noopener">5a08a55</a> (the commit meant for the issueurl branch)</strong>. This procedure overwrites commit history, and that is looked down upon by some, but hey, I doubt the “masters” of <em>linkchecker</em> would merge <a href="https://github.com/linkchecker/linkchecker/pull/176" rel="nofollow noopener">#176</a> (with a reverted <a href="https://github.com/linkchecker/linkchecker/pull/176/commits/1087744051dd4ef6f67501e81162667e7777d191" rel="nofollow noopener">1087744</a>) and then accept <a href="https://github.com/linkchecker/linkchecker/pull/175" rel="nofollow noopener">#175</a>. I think one should keep things clean when presenting work.</p>
<p>I hope that helps! <em>Git</em> it together, won’t ya :wink:</p>

**Comment by regexaurus** [comment]
<p><a class="mention" href="/imasaru">@imasaru</a></p>
<p>I tried your recommendation:</p>
<p>$ git branch<br>
* issueurl-regexaurus<br>
master<br>
typos-regexaurus</p>
<p>$ git reset --hard</p>
<p>$ git status<br>
On branch issueurl-regexaurus<br>
nothing to commit, working directory clean</p>
<p>$ git cherry-pick 5a08a55<br>
On branch issueurl-regexaurus<br>
You are currently cherry-picking commit 5a08a55.</p>
<p>nothing to commit, working directory clean<br>
The previous cherry-pick is now empty, possibly due to conflict resolution.<br>
If you wish to commit it anyway, use:</p>
<p>git commit --allow-empty</p>
<p>Otherwise, please use ‘git reset’</p>
<p>$ git commit --allow-empty<br>
[issueurl-regexaurus 48a0247] Updated support (issues) URL</p>
<p>At this point, CONTRIBUTING.mdwn reflected the typo correction from 1087744. I wasn’t sure this would do what I wanted. Yes, I think you’re correct that I mistakenly created <em>issueurl-regexaurus</em> from <em>typos-regexaurus</em>. I ended up commenting on and closing PR <span class="hashtag">#176</span> and deleted <em>issueurl-regexaurus</em>, then created a new <em>issueurl-regexaurus</em> from <em>master</em>.</p>
<p>Hey, I’m trying, I’m trying. :smileyvery-happy:</p>

**Comment by imasaru** [comment]
<p>Hmm, it looks like you reset to where the <em>issueurl</em> branch is already at, hence the cherry-pick was empty because the changes were already made.</p>
<p>You should try:  <strong>git reset --hard master</strong> (because your master branch seems to be even with the parent, so that should do the trick) and then do the cherry-pick. Once you have picked, then force-push to origin (GitHub):  <strong>git push origin issueurl-regexaurus -f</strong> , which will (f)orcibly push your commit to GitHub to overwrite the current history.</p>
